### PR TITLE
Add Random number to N3

### PR DESCRIPTION
**Pull request by Liaojinghui**
Close pull request #590

### CODE REVIEW

File: src/DBFTPlugin/ConsensusContext.cs
Comment (ZhangTao1596) [inline src/DBFTPlugin/ConsensusContext.cs:423]: You should use the proof in `RrepareRequest` instead creating a new one. Proof should be caculated from Primary private key.

Comment (nicolegys) [reply src/DBFTPlugin/ConsensusContext.cs:423]: Yes, creating a new one will cause issue when stopping the CNs at different stages. They can no longer achieve the consensus.
image

Comment (Liaojinghui) [reply src/DBFTPlugin/ConsensusContext.cs:423]: > You should use the proof in `RrepareRequest` instead creating a new one. Proof should be caculated from Primary private key.

I did not get that, isn't that only the primary could send the recovery message?

File: src/DBFTPlugin/Consensus/ConsensusContext.cs
Comment (nicolegys) [inline src/DBFTPlugin/Consensus/ConsensusContext.cs:208]: `nonce` needs to be reset, too

**Comment by nicolegys** [comment]
Test passed, waiting for review.

**Comment by shargon** [comment]
Too many consensus changes, could you split the refactor and the random number in two different pr?

**Comment by Liaojinghui** [comment]
> Too many consensus changes, could you split the refactor and the random number in two different pr?

I also thought about create another pr, but if so, I will either have to wait until random is merged or face tons of conflicts afterward. Since I just make big class into smaller partial files and did not change any code, it will not cause any issue to the logic.

**Review by Qiao-Jin — DISMISSED** [review]

**Comment by Qiao-Jin** [comment]
Please correct format

File: src/DBFTPlugin/DBFTPlugin.csproj
Comment (shargon) [inline src/DBFTPlugin/DBFTPlugin.csproj:18]: If we add this we should add it to all the projects, I think that it can be removed

Comment (Liaojinghui) [reply src/DBFTPlugin/DBFTPlugin.csproj:18]: lol, i just think this would save a lot of work in format updating ( or is it just me who make a lot of format errors?).  it wont hurt even if we add it to all projects, right?

Comment (shargon) [reply src/DBFTPlugin/DBFTPlugin.csproj:18]: I think that it's not a bad change, but seems weird if only in one project was added, isn't it?

Comment (superboyiii) [reply src/DBFTPlugin/DBFTPlugin.csproj:18]: > @shargon lol, i just think this would save a lot of work in format updating ( or is it just me who make a lot of format errors?). it wont hurt even if we add it to all projects, right?

I think you could add to the whole project.

**Comment by superboyiii** [comment]
https://github.com/Liaojinghui/neo-modules/blob/f12250decb632e57aec04522217a699e28ddc424/src/Directory.Build.props#L18
Please change to `3.0.0-CI01285`. @Liaojinghui

**Comment by Qiao-Jin** [comment]
Ping

**Comment by superboyiii** [comment]
@shargon Pls review again.

File: src/ApplicationLogs/ApplicationLogs.csproj
Comment (shargon) [inline src/ApplicationLogs/ApplicationLogs.csproj:8]: It could be done in `Directory.Build.props`?

Comment (Liaojinghui) [reply src/ApplicationLogs/ApplicationLogs.csproj:8]: > It could be done in `Directory.Build.props`?

Well, let's find out

Comment (Liaojinghui) [reply src/ApplicationLogs/ApplicationLogs.csproj:8]: @shargon unfortunately, I did not figure out a way to do that.

**Comment by superboyiii** [comment]
@shargon Pls review again.

**Review by superboyiii — APPROVED** [review]

**Review by erikzhang — APPROVED** [review]

**Pull request merged**
